File: cpp/assert/assert_arguments_length.cpp
// Make sure JS function was called with correct amount of arguments

void assert_arguments_length(const v8::FunctionCallbackInfo < v8::Value > &aArguments, size_t aMinLength, size_t aMaxLength, const char* aFunctionName) {
    // Make sure JS function was called with correct amount of arguments
    size_t l = aArguments.Length();
    if (l < aMinLength || l > aMaxLength) {
        hint_args(aFunctionName, aArguments);
        if (aMinLength == aMaxLength) {
            fatal_msg_stack(aArguments.GetIsolate(), EXIT_ARGS, "%s require exactly %d arguments but %d was given", aFunctionName, aMinLength, l);
            exit(EXIT_UNREACHABLE);
        } else if (l < aMinLength) {
            fatal_msg_stack(aArguments.GetIsolate(), EXIT_ARGS, "%s require at least %d arguments but only %d was given", aFunctionName, aMinLength, l);
            exit(EXIT_UNREACHABLE);
        } else if (l > aMaxLength) {
            fatal_msg_stack(aArguments.GetIsolate(), EXIT_ARGS, "%s require at most %d arguments but %d was given", aFunctionName, aMaxLength, l);
            exit(EXIT_UNREACHABLE);
        }
    }
}

File: js/assert/assert_arguments_length.js
// Assert that arguments have correct number of items
// linter: ngspicejs-lint --internal
"use strict";

function assert_arguments_length(aArguments, aMinLength, aMaxLength, aFunctionName) {
    // Assert that arguments have correct number of items
    var al = aArguments.length;
    if (al >= aMinLength && al <= aMaxLength) {
        return;
    }
    if (aArguments && (aArguments[0] === api)) {
        throw new Exception('Function "' + aFunctionName + '" requires ' + (aMinLength === aMaxLength ? aMinLength : (aMinLength + " to " + aMaxLength)) + ' arguments');
    }
    if (aMinLength !== undefined && aMinLength === aMaxLength && al !== aMinLength) {
        hint_args(aFunctionName, aArguments);
        throw new Exception(aFunctionName + " requires exactly " + aMinLength + " arguments but " + al + " was given");
    }
    if (aMinLength !== undefined && al < aMinLength) {
        hint_args(aFunctionName, aArguments);
        throw new Exception(aFunctionName + " requires at least " + aMinLength + " arguments but " + al + " was given");
    }
    if (aMaxLength !== undefined && al > aMaxLength) {
        hint_args(aFunctionName, aArguments);
        throw new Exception(aFunctionName + " requires at most " + aMaxLength + " arguments but " + al + " was given");
    }
}

globalThis.exports = {assert_arguments_length};
Internal.assert_arguments_length = assert_arguments_length;
